components/datetime/src/format/input.rs

```diff
@@ -17,24 +18,106 @@ use icu_time::{zone::UtcOffset, Time, TimeZone};
 
-#[derive(Debug, Copy, Clone)]
-pub(crate) struct ExtractedInput {
+#[cfg(doc)]
+use crate::input::*;
+
+/// An input bag with all possible datetime input fields.
+///
+/// Each input field may or may not be required, depending on the field set
+/// and the options.
+#[derive(Debug, Copy, Clone, Default)]
+#[non_exhaustive]
+pub struct DateTimeInputUnchecked {
+    /// The year, required for field sets with years (`Y`).
     pub(crate) year: Option<YearInfo>,
+    /// The month, required for field sets with months (`M`)
     pub(crate) month: Option<MonthInfo>,
+    /// The day-of-month, required for field sets with days (`D`).
     pub(crate) day_of_month: Option<DayOfMonth>,
+    /// The weekday, required for field sets with weekdays (`E`).
     pub(crate) iso_weekday: Option<Weekday>,
+    /// The day-of-year, required for field sets with weeks.
     pub(crate) day_of_year: Option<DayOfYearInfo>,
+    /// The hour, required for field sets with times (`T`).
     pub(crate) hour: Option<Hour>,
+    /// The minute, required for field sets with times (`T`).
     pub(crate) minute: Option<Minute>,
+    /// The second, required for field sets with times (`T`).
     pub(crate) second: Option<Second>,
+    /// The subsecond, required for field sets with times (`T`).
     pub(crate) subsecond: Option<Nanosecond>,
+    /// The time zone ID, required for field sets with
+    /// certain time zone styles.
     pub(crate) time_zone_id: Option<TimeZone>,
+    /// The time zone UTC offset, required for field sets with
+    /// certain time zone styles.
     pub(crate) offset: Option<UtcOffset>,
+    /// The time zone variant, required for field sets with
+    /// certain time zone styles.
     pub(crate) zone_variant: Option<TimeZoneVariant>,
+    /// The local ISO time, required for field sets with
+    /// certain time zone styles.
     pub(crate) local_time: Option<(Date<Iso>, Time)>,
 }
 
-impl ExtractedInput {
+macro_rules! set_field {
+    ($type:ident as $trait:ident, $receiver:expr, $assoc_type:ident, $input:expr) => {
+        $receiver = GetField::<<$type as $trait>::$assoc_type>::get_field($input).into_option()
+    };
+    (@date, $receiver:expr, $assoc_type:ident, $input:expr) => {
+        set_field!(
+            DateFieldSet as DateInputMarkers,
+            $receiver,
+            $assoc_type,
+            $input
+        )
+    };
+    (@time, $receiver:expr, $assoc_type:ident, $input:expr) => {
+        set_field!(TimeFieldSet as TimeMarkers, $receiver, $assoc_type, $input)
+    };
+    (@zone, $receiver:expr, $assoc_type:ident, $input:expr) => {
+        set_field!(ZoneFieldSet as ZoneMarkers, $receiver, $assoc_type, $input)
+    };
+}
+
+impl DateTimeInputUnchecked {
+    /// Sets all fields from a [`Date`] input.
+    pub fn set_date_fields<C: Calendar, A: AsCalendar<Calendar = C>>(&mut self, input: Date<A>) {
+        set_field!(@date, self.year, YearInput, &input);
```

Comment:
Can you *please* just inline these. They are completely unreadable behind multiple layers of macros, and they would just be one liners

Reply:
still better

Reply:
done